ddtrace/internal/datadog/profiling/dd_wrapper/src/global_cache.cpp

```diff
@@ -0,0 +1,52 @@
+#include "global_cache.hpp"
+
+#include <mutex>
+#include <stdexcept>
+
+using namespace Datadog;
+
+GlobalCache&
+GlobalCache::get_singleton()
+{
+    static GlobalCache instance;
+    return instance;
+}
+
+Sample&
+GlobalCache::get(std::thread::id id)
+{
+    GlobalCache& global_store = get_singleton();
+
+    std::unique_lock<std::mutex> lock(global_store.sample_storage_mtx);
+    auto& samples = global_store.sample_cache;
+    auto it = samples.find(id); // Attempt to find the sample once
+
+    if (it == samples.end()) {
+        // Sample not found, unlock the mutex while creating the sample
```

Comment:
is the issue that the `sample_builder` gets a lock and is not reentrant?

Reply:
The main issue was just performance.  Creating a sample takes time, so reduce the lock surface.  This is somewhat silly because bouncing all over the lock is arguably worse than just holding it for some additional amount of time, but since this is a one-time call per thread, maybe not too important.

Either way, I removed all this code.  I don't think we need it.  Added comments in the .md